src/dhcpmon/src/dhcp_device.c

```diff
@@ -0,0 +1,456 @@
+    case 3:
+        context->counters[dir].request++;
+        if ((context->is_uplink && dir == DHCP_TX) || (!context->is_uplink && dir == DHCP_RX)) {
+            __atomic_fetch_add(&glob_counters[dir].request, 1, __ATOMIC_SEQ_CST);
+        }
+        break;
+    case 5:
+        context->counters[dir].ack++;
+        if ((!context->is_uplink && dir == DHCP_TX) || (context->is_uplink && dir == DHCP_RX)) {
+            __atomic_fetch_add(&glob_counters[dir].ack, 1, __ATOMIC_SEQ_CST);
+        }
+        break;
+    case 4: // type: Decline
+    case 6 ... 8:
+        // type: NAK, Release, Inform
+        break;
+    default:
+        syslog(LOG_WARNING, "handle_dhcp_option_53(%s): Unknown DHCP option 53 type %d", context->intf, dhcp_option[2]);
+        break;
+    }
+}
+
+/**
+ * @code read_callback(fd, event, arg);
+ *
+ * @brief callback for libpcap which is called every time out in order to process queued packet capture
+ *
+ * @param fd            socket to read from
+ * @param event         libevent triggered event
+ * @param arg           user provided argument for callback (interface context)
+ * @param packet        pointer to packet data
+ *
+ * @return none
+ */
+static void read_callback(int fd, short event, void *arg)
+{
+    dhcp_device_context_t *context = (dhcp_device_context_t*) arg;
+    ssize_t buffer_sz;
+
+    while ((event == EV_READ) &&
+           ((buffer_sz = recv(fd, context->buffer, context->snaplen, MSG_DONTWAIT)) > 0)) {
+        struct ether_header *ethhdr = (struct ether_header*) context->buffer;
+        struct udphdr *udp = (struct udphdr*) (context->buffer + UDP_START_OFFSET);
+        int dhcp_option_offset = DHCP_START_OFFSET + DHCP_OPTIONS_HEADER_SIZE;
+
+        if ((buffer_sz > UDP_START_OFFSET + sizeof(struct udphdr) + DHCP_OPTIONS_HEADER_SIZE) &&
+            (ntohs(udp->len) > DHCP_OPTIONS_HEADER_SIZE)) {
+            int dhcp_sz = ntohs(udp->len) < buffer_sz - UDP_START_OFFSET - sizeof(struct udphdr) ?
+                          ntohs(udp->len) : buffer_sz - UDP_START_OFFSET - sizeof(struct udphdr);
+            int dhcp_option_sz = dhcp_sz - DHCP_OPTIONS_HEADER_SIZE;
+            const u_char *dhcp_option = context->buffer + dhcp_option_offset;
+            dhcp_packet_direction_t dir = (ethhdr->ether_shost[0] == context->mac[0] &&
+                                           ethhdr->ether_shost[1] == context->mac[1] &&
+                                           ethhdr->ether_shost[2] == context->mac[2] &&
+                                           ethhdr->ether_shost[3] == context->mac[3] &&
+                                           ethhdr->ether_shost[4] == context->mac[4] &&
+                                           ethhdr->ether_shost[5] == context->mac[5]) ?
+                                           DHCP_TX : DHCP_RX;
+            int offset = 0;
+            int stop_dhcp_processing = 0;
+            while ((offset < (dhcp_option_sz + 1) )&& dhcp_option[offset] != 255) {
```

Comment:
remove space between parenthesis

Reply:
And add space after them (between the closing parenthesis and the `&&`) -- looks like the space was accidentally added in the wrong spot.

Reply:
Yupe, fixed.